packages/create-project/src/helpers/install.ts

```diff
@@ -0,0 +1,108 @@
+import chalk from 'chalk';
+import spawn from 'cross-spawn';
+
+interface InstallArgs {
+  /**
+   * Indicate whether to install packages using Yarn.
+   */
+  useYarn: boolean;
+  /**
+   * Indicate whether there is an active Internet connection.
+   */
+  isOnline: boolean;
+  /**
+   * Indicate whether the given dependencies are devDependencies.
+   */
+  devDependencies?: boolean;
+}
+
+/**
+ * Spawn a package manager installation with either Yarn or NPM.
+ *
+ * @returns A Promise that resolves once the installation is finished.
+ */
+export function install(
+  root: string,
+  dependencies: string[] | null,
+  { useYarn, isOnline, devDependencies }: InstallArgs
+): Promise<void> {
+  /**
+   * NPM-specific command-line flags.
+   */
+  const npmFlags: string[] = [];
+  /**
+   * Yarn-specific command-line flags.
+   */
+  const yarnFlags: string[] = [];
+  /**
+   * Return a Promise that resolves once the installation is finished.
+   */
+  return new Promise((resolve, reject) => {
+    let args: string[];
+    const command: string = useYarn ? 'yarnpkg' : 'npm';
+
+    if (dependencies && dependencies.length) {
+      /**
+       * If there are dependencies, run a variation of `{displayCommand} add`.
+       */
+      if (useYarn) {
+        /**
+         * Call `yarn add --exact (--offline)? (-D)? ...`.
+         */
+        args = ['add', '--exact'];
+        if (!isOnline) args.push('--offline');
+        args.push('--cwd', root);
+        if (devDependencies) args.push('--dev');
+        args.push(...dependencies);
+      } else {
+        /**
+         * Call `npm install [--save|--save-dev] ...`.
+         */
+        args = ['install', '--save-exact'];
+        args.push(devDependencies ? '--save-dev' : '--save');
+        args.push(...dependencies);
+      }
+    } else {
+      /**
+       * If there are no dependencies, run a variation of `{displayCommand}
+       * install`.
+       */
+      args = ['install'];
+      if (useYarn) {
+        if (!isOnline) {
```

Comment:
Why do we bother testing if the user is online?
Does this provide better DX over just letting npm or yarn detect that?

Reply:
Yes, we can provide a clearer error message—npm just hangs with an empty progress bar. Yarn is better but could be clearer.

![image](https://user-images.githubusercontent.com/251288/132248485-9d85bd6d-6d79-4988-a062-6450f8a7f791.png)